### PR TITLE
[DNM][stdlib] Add reduce(_:) method for Sequence to standard library

### PR DESCRIPTION
**Pull request by erica**
Introduces a folding operator that seeds itself from the elements
passed to reduce, removing the need for an explicit value seed.

Implements [SE-nnnnnnnn](https://github.com/apple/swift-evolution/blob/master/proposals/nnnnnnnn-fold.md)

* [Swift Evolution thread](https://forums.swift.org/t/pitch-reducing-a-sequence-onto-its-own-elements-to-simplify-code-and-reduce-errors/19243/52)

### CODE REVIEW

**Comment by erica** [comment]
@swift-ci please test source compatibility

**Comment by erica** [comment]
@swift-ci please test

**Comment by erica** [comment]
@swift-ci please benchmark

File: stdlib/public/core/SequenceAlgorithms.swift [hunk 1]
```diff
@@ -677,6 +677,80 @@ extension Sequence {
...
+    (_ partialResult: Element, _ current: Element) throws -> Element
+  ) rethrows -> Element? {
+      var iterator = makeIterator()
+      guard var accumulator = iterator.next() else {
+        return nil
+      }
+      while let element = iterator.next() {
+        accumulator = try nextPartialResult(accumulator, element)
+      }
+      return accumulator
```
Comment (Azoy) [inline stdlib/public/core/SequenceAlgorithms.swift]: Nit: 2 space tab for the body

Comment (erica) [reply stdlib/public/core/SequenceAlgorithms.swift]: Darn you copy/paste into Xcode! Amended.

**Comment by erica** [comment]
@swift-ci please test

**Comment by erica** [comment]
@swift-ci please test

**Comment by erica** [comment]
@swift-ci please smoketest

**Comment by erica** [comment]
@swift-ci please smoketest

**Comment by millenomi** [comment]
@swift-ci please smoke test

File: stdlib/public/core/SequenceAlgorithms.swift [hunk 2]
```diff
@@ -677,6 +677,161 @@ extension Sequence {
...
+  /// 2. The closure is called repeatedly with the current partial result
+  ///    and each successive member of the sequence
+  /// 3. When the sequence is exhausted, the method returns the last value
+  ///    established by the closure.
+  ///
+  /// If the sequence has no elements, `reduce` returns `nil`.
+  ///
+  /// If the sequence has one element, `reduce` returns that element.
+  ///
+  /// This method uses an (inout Element, Element) -> Element closure.
```
Comment (Dante-Broggi) [inline stdlib/public/core/SequenceAlgorithms.swift]:
```suggestion
  /// This method uses an (inout Element, Element) -> Void closure.
```

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

**Pull request closed without merging**